torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,801 @@
+    v_weight=0.1,
+    train_device="cpu",
+    actor_device="cpu",
+    reanalyze_device="cpu",
+    reanalyze_workers=0,
+)
+
+
+class MuZeroAgentAdapter(AgentAdapter):
+    def tensor_from_cog_obs(self, cog_obs, device=None):
+        pb_array = cog_obs.vectorized
+        np_array = np.frombuffer(pb_array.data, dtype=pb_array.dtype).reshape(*pb_array.shape)
+        return torch.tensor(np_array, dtype=self._dtype, device=device)
+
+    @staticmethod
+    def decode_cog_policy_value(cog_action):
+        policy = np.frombuffer(cog_action.policy.data, dtype=cog_action.policy.dtype).reshape(*cog_action.policy.shape)
+        value = cog_action.value
+        return policy, value
+
+    @staticmethod
+    def decode_cog_action(cog_action):
+        action = cog_action.discrete_action
+        return action
+
+    def __init__(self):
+        super().__init__()
+        self._model_cache = LRU(2)  # memory issue?
+        self._dtype = torch.float
+
+    def _create(self, model_id, *, obs_dim, act_dim, device, training_config):
+        return MuZeroAgent(obs_dim=obs_dim, act_dim=act_dim, device=device, training_config=training_config)
+
+    def _load(self, model_id, version_number, version_user_data, model_data_f):
+        return MuZeroAgent.load(model_data_f, "cpu")
+
+    def _save(self, model, model_data_f):
+        assert isinstance(model, MuZeroAgent)
+        model.save(model_data_f)
+        return {}
+
+    def _create_actor_implementations(self):
+        async def _single_agent_muzero_actor_implementation(actor_session):
+            actor_session.start()
+            agent, _ = await self.retrieve_version(actor_session.config.model_id, -1)
+            agent = copy.deepcopy(agent)
+            agent.set_device(actor_session.config.device)
+
+            worker = AgentTrialWorker(agent, actor_session.config)
+            worker.start()
+
+            try:
+                async for event in actor_session.event_loop():
+                    assert worker.is_alive()
+                    if event.observation and event.type == cogment.EventType.ACTIVE:
+                        await worker.put_event(event)
+                        action = await worker.get_action()
+                        actor_session.do_action(action)
+            finally:
+                worker.terminate()
+
+        return {
+            "muzero_mlp": (_single_agent_muzero_actor_implementation, ["agent"]),
+        }
+
+    def _create_run_implementations(self):
+        """
+        Create all the available run implementation for this adapter
+        Returns:
+            dict[impl_name: string, (sample_producer_impl: Callable, run_impl: Callable, default_run_config)]: key/value definition for the available run implementations.
+        """
+
+        async def _single_agent_muzero_sample_producer_implementation(run_sample_producer_session):
+            return await single_agent_muzero_sample_producer_implementation(self, run_sample_producer_session)
```

Comment:
In fact those could be methods of the class, you wouldn't need this indirection.

Reply:
Done